modules/javafx.graphics/src/main/java/javafx/scene/layout/HeaderBar.java

```diff
@@ -0,0 +1,523 @@
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package javafx.scene.layout;
+
+import com.sun.javafx.geom.Vec2d;
+import javafx.beans.property.ObjectProperty;
+import javafx.beans.property.ObjectPropertyBase;
+import javafx.css.StyleableDoubleProperty;
+import javafx.geometry.HPos;
+import javafx.geometry.Insets;
+import javafx.geometry.NodeOrientation;
+import javafx.geometry.Orientation;
+import javafx.geometry.Pos;
+import javafx.geometry.VPos;
+import javafx.scene.Node;
+import javafx.stage.StageStyle;
+
+/**
+ * A client-area header bar that is used as a replacement for the system-provided header bar in stages
+ * with the {@link StageStyle#EXTENDED} style. This class enables the <em>click-and-drag</em> and
+ * <em>double-click to maximize</em> behaviors that are usually afforded by system-provided header bars.
+ * The entire {@code HeaderBar} background is draggable by default, but its content is not. Applications
+ * can specify draggable content nodes of the {@code HeaderBar} with the {@link #setDraggable} method.
+ * <p>
+ * {@code HeaderBar} is a layout container that allows applications to place scene graph nodes
+ * in three areas: {@link #leadingProperty() leading}, {@link #centerProperty() center}, and
+ * {@link #trailingProperty() trailing}. {@code HeaderBar} ensures that the leading and trailing areas
+ * account for the default window buttons (minimize, maximize, close). If a child is configured to be
+ * centered in the {@code center} area, it is laid out with respect to the stage, and not with respect
+ * to the {@code center} area. This ensures that the child will appear centered in the stage regardless
+ * of leading or trailing children or the platform-specific placement of default window buttons.
+ * <p>
+ * All children will be resized to their preferred widths and extend the height of the {@code HeaderBar}.
+ * {@code HeaderBar} honors the minimum, preferred, and maximum sizes of its children. If the child's
+ * resizable range prevents it from be resized to fit within its position, it will be vertically centered
+ * relative to the available space; this alignment can be customized with a layout constraint.
```

Comment:
what if the components are too large to fit, will clipping occur?  should this behavior be mentioned?

Reply:
The components can't be too large to fit:
`{@code HeaderBar} honors the minimum, preferred, and maximum sizes of its children.`

Reply:
it is probably ok - I am curious what @kevinrushforth might think.

it might be either obvious that the window will cut off the HB if it is wider than the window width, or it should be explicitly mentioned (there is probably enough text to go through to kind of understand the behavior in the case of over-constrained HB, or the dev can just try it).